infer/src/backend/dom.ml

```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
   in
   res
 
+module MaxCliqueState = Caml.Set.Make (struct
+  type t = int list (* invariant: increasing *)
+  let rec compare xs ys = match xs, ys with
+    | [], [] -> 0
+    | [], _ -> -1
+    | _, [] -> +1
+    | x ::  _, y ::  _ when x < y -> -1
+    | x ::  _, y ::  _ when x > y -> +1
+    | _ :: xs, _ :: ys -> compare xs ys
+end)
+
+let rec filter_to n predicate xs x =
+  if x >= n then List.rev xs
+  else if predicate x xs then filter_to n predicate (x :: xs) (x + 1)
+  else filter_to n predicate xs (x + 1)
+
+(* Given a graph, initialize a maxclique generator. The vertices of the
+(undirected) graph are the integers 0,...,n-1, and its edges are given by the
+function [edge]. *)
+let initgen_maxclique (n : int) (edge : int -> int -> bool) =
+  let compatible x xs = List.for_all xs ~f:(edge x) in
+  let first = filter_to n compatible [] 0 in
+  let module S = MaxCliqueState in
+  (S.add first S.empty, n, edge)
+
+(* Given a maxclique generator, returns a pair (xs, g) with the next maxclique
+[xs] and the updated generator state [g]. If no more maxclique, raises
+[Not_found]. For the algorithm, see the JYP paper mentioned below. *)
+let next_maxclique (queue, n, edge) =
+  let module S = MaxCliqueState in
+  let xs = S.min_elt queue in (* raises [Not_found] at the end *)
+  let queue = S.remove xs queue in
+  let ys =
+    let ok_yx y x = x < y && not (edge x y) in
+    let ok_y y _ = List.exists xs ~f:(ok_yx y) in
+    filter_to n ok_y [] 0 in
+  let do_y queue y =
+    let xy_compat x = x < y && edge x y in
+    let zs = List.filter xs ~f:xy_compat @ [y] in
+    let maxclique zs = (* is zs a maxclique for the subgraph induced by 0..y? *)
+      let rec loop u vs = u > y || (match u, vs with
+        | u, v :: vs when Int.equal u v -> loop (u + 1) vs
+        | u, vs -> not (List.for_all zs ~f:(edge u)) && loop (u + 1) vs) in
+      loop 0 zs in
+    let fillup zs = (* greedily add vertices from y+1..n-1 to clique *)
+      let compatible z zs = List.for_all zs ~f:(edge z) in
+      filter_to n compatible (List.rev zs) (y + 1) in
+    if maxclique zs then S.add (fillup zs) queue else queue in
+  let queue = List.fold ~init:queue ~f:do_y ys in
```

Comment:
I'm not sure I get this:
They _could_, but it seems to me it would be quite ugly. These two loops appear separated even in the pseudocode from the JYP paper.